test/sharness/t0160-resolve.sh

```diff
@@ -14,104 +14,141 @@ test_expect_success "resolve: prepare files" '
+test_resolve "/ipns/$id_hash/b" "/ipfs/$b_hash"
+test_resolve "/ipns/$id_hash/b/c" "/ipfs/$c_hash"
+
+test_name_publish "/ipfs/$b_hash" --ttl=0s
+test_name_resolve "/ipfs/$b_hash"
+test_resolve "/ipns/$id_hash" "/ipfs/$b_hash"
+test_resolve "/ipns/$id_hash/c" "/ipfs/$c_hash"
+
+test_name_publish "/ipfs/$c_hash" --ttl=0s
+test_name_resolve "/ipfs/$c_hash"
+test_resolve "/ipns/$id_hash" "/ipfs/$c_hash"
+
+# The following tests test the caching by publishing a new name and expecting
+# the previous resolve result to stay cached.  The first test_(name_)resolve
+# after a test_name_publish will generate the cache entry and any subsequent
+# test_resolve invocations will use that cache entry until the expiry timer
+# finishes (test_timer_wait).
+
+TTL=10
+
+test_name_publish "/ipfs/$a_hash" --ttl="${TTL}s"
+test_name_resolve "/ipfs/$a_hash"
+
+# The cache entry has expired when this finishes.
+test_timer_start EXPIRY_TIMER "${TTL}s"
+# Publish a new version now, the previous version should still be cached.
+test_name_publish "/ipfs/$b_hash" --ttl="${TTL}s"
+
+test_resolve "/ipns/$id_hash" "/ipfs/$a_hash"
+test_resolve "/ipns/$id_hash/b" "/ipfs/$b_hash"
+test_resolve "/ipns/$id_hash/b/c" "/ipfs/$c_hash"
+
+# Make sure the expiry timer is still running, otherwise the result might be
+# wrong.  If this fails, we will need to increase TTL above to give enough time
+# for the tests.
+test_expect_success "tests did not take too long" '
+  test_timer_is_running "$EXPIRY_TIMER"
+'
+test_expect_success "previous version is no longer cached" '
+  test_timer_wait "$EXPIRY_TIMER"
+'
 
-test_resolve_cmd_fail() {
-	test_resolve "/ipfs/$a_hash" "/ipfs/$a_hash"
-	test_resolve "/ipfs/$a_hash/b" "/ipfs/$b_hash"
-	test_resolve "/ipfs/$a_hash/b/c" "/ipfs/$c_hash"
-	test_resolve "/ipfs/$b_hash/c" "/ipfs/$c_hash"
+test_name_resolve "/ipfs/$b_hash"
 
-	test_resolve_setup_name_fail "/ipfs/$a_hash"
-	test_resolve_fail "/ipns/$id_hash" "/ipfs/$a_hash"
-	test_resolve_fail "/ipns/$id_hash/b" "/ipfs/$b_hash"
-	test_resolve_fail "/ipns/$id_hash/b/c" "/ipfs/$c_hash"
+test_timer_start EXPIRY_TIMER "${TTL}s"
+# Publish a new version now, the previous version should still be cached.
+test_name_publish "/ipfs/$c_hash" # Default TTL for the final one.
 
-	test_resolve_setup_name_fail "/ipfs/$b_hash"
-	test_resolve_fail "/ipns/$id_hash" "/ipfs/$b_hash"
-	test_resolve_fail "/ipns/$id_hash/c" "/ipfs/$c_hash"
+test_resolve "/ipns/$id_hash" "/ipfs/$b_hash"
+test_resolve "/ipns/$id_hash/c" "/ipfs/$c_hash"
 
-	test_resolve_setup_name_fail "/ipfs/$c_hash"
-	test_resolve_fail "/ipns/$id_hash" "/ipfs/$c_hash"
-}
+test_expect_success "tests did not take too long" '
+  test_timer_is_running "$EXPIRY_TIMER"
+'
+test_expect_success "previous version is no longer cached" '
+  test_timer_wait "$EXPIRY_TIMER"
+'
 
-# should work offline
-test_resolve_cmd
+test_name_resolve "/ipfs/$c_hash"
+test_resolve "/ipns/$id_hash" "/ipfs/$c_hash"
 
-# should work online
-test_launch_ipfs_daemon
```

Comment:
Those lines explicitly test the caching done by the daemon in online mode. I can put them into a function but they will fail if executed while offline because there is no daemon to cache resolve results.